// nodejs-polars/bin/datatypes/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.polarsTypeToConstructor = exports.DTYPE_TO_FFINAME = exports.Field = exports.TimeUnit = exports.DataType = void 0;
const datatype_1 = require("./datatype");
Object.defineProperty(exports, "DataType", { enumerable: true, get: function () { return datatype_1.DataType; } });
Object.defineProperty(exports, "TimeUnit", { enumerable: true, get: function () { return datatype_1.TimeUnit; } });
var field_1 = require("./field");
Object.defineProperty(exports, "Field", { enumerable: true, get: function () { return field_1.Field; } });
const polars_internal_1 = __importDefault(require("../internals/polars_internal"));
/**
 * @ignore
 */
exports.DTYPE_TO_FFINAME = {
    Int8: "I8",
    Int16: "I16",
    Int32: "I32",
    Int64: "I64",
    UInt8: "U8",
    UInt16: "U16",
    UInt32: "U32",
    UInt64: "U64",
    Float32: "F32",
    Float64: "F64",
    Bool: "Bool",
    Utf8: "Str",
    List: "List",
    Date: "Date",
    Datetime: "Datetime",
    Time: "Time",
    Object: "Object",
    Categorical: "Categorical",
    Struct: "Struct",
};
const POLARS_TYPE_TO_CONSTRUCTOR = {
    Float32(name, values, strict) {
        return polars_internal_1.default.JsSeries.newOptF64(name, values, strict);
    },
    Float64(name, values, strict) {
        return polars_internal_1.default.JsSeries.newOptF64(name, values, strict);
    },
    Int8(name, values, strict) {
        return polars_internal_1.default.JsSeries.newOptI32(name, values, strict);
    },
    Int16(name, values, strict) {
        return polars_internal_1.default.JsSeries.newOptI32(name, values, strict);
    },
    Int32(name, values, strict) {
        return polars_internal_1.default.JsSeries.newOptI32(name, values, strict);
    },
    Int64(name, values, strict) {
        return polars_internal_1.default.JsSeries.newOptI64(name, values, strict);
    },
    UInt8(name, values, strict) {
        return polars_internal_1.default.JsSeries.newOptU32(name, values, strict);
    },
    UInt16(name, values, strict) {
        return polars_internal_1.default.JsSeries.newOptU32(name, values, strict);
    },
    UInt32(name, values, strict) {
        return polars_internal_1.default.JsSeries.newOptU32(name, values, strict);
    },
    UInt64(name, values, strict) {
        return polars_internal_1.default.JsSeries.newOptU64(name, values, strict);
    },
    Date(name, values, strict) {
        return polars_internal_1.default.JsSeries.newOptI64(name, values, strict);
    },
    Datetime(name, values, strict) {
        return polars_internal_1.default.JsSeries.newOptI64(name, values, strict);
    },
    Bool(name, values, strict) {
        return polars_internal_1.default.JsSeries.newOptBool(name, values, strict);
    },
    Utf8(name, values, strict) {
        return polars_internal_1.default.JsSeries.newOptStr(name, values, strict);
    },
    Categorical(name, values, strict) {
        return polars_internal_1.default.JsSeries.newOptStr(name, values, strict);
    },
    List(name, values, _strict, dtype) {
        return polars_internal_1.default.JsSeries.newList(name, values, dtype);
    },
};
/** @ignore */
const polarsTypeToConstructor = (dtype) => {
    const ctor = POLARS_TYPE_TO_CONSTRUCTOR[dtype.variant];
    if (!ctor) {
        throw new Error(`Cannot construct Series for type ${dtype.variant}.`);
    }
    return ctor;
};
exports.polarsTypeToConstructor = polarsTypeToConstructor;
