// nodejs-polars/bin/cfg.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Config = void 0;
const native_polars_1 = __importDefault(require("./native-polars"));
/**
 * @ignore
 */
exports.Config = {
    setUtf8Tables() {
        process.env["POLARS_FMT_NO_UTF8"] = undefined;
        return this;
    },
    setAsciiTables() {
        process.env["POLARS_FMT_NO_UTF8"] = "1";
        return this;
    },
    setTblWidthChars(width) {
        process.env["POLARS_TABLE_WIDTH"] = String(width);
        return this;
    },
    setTblRows(n) {
        process.env["POLARS_FMT_MAX_ROWS"] = String(n);
        return this;
    },
    setTblCols(n) {
        process.env["POLARS_FMT_MAX_COLS"] = String(n);
        return this;
    },
    setGlobalStringCache() {
        native_polars_1.default.toggleStringCache(true);
        return this;
    },
    unsetGlobalStringCache() {
        native_polars_1.default.toggleStringCache(false);
        return this;
    },
};
